### DISCUSSION TITLE
How to append content of a .qmd file to the one currently rendered using a lua fragment?

### DISCUSSION THREAD

**Discussion by alesvomacka** [quarto-dev/quarto-cli#6025] [Q&A] [answered]
### Description

Hi,
I'm trying to create a simple lua script that would to append content of one .qmd to another. The goal is to automatically append a slide containing contact info at the end of my presentations without having to do so manually.

Currently, the script looks like this:

```lua
-- End slide content
local end_slide = io.open("ending.qmd", "r")
local end_content = end_slide:read("*all")
end_slide:close()

-- Append to the rest of the slides
local main_slides = io.open("template.qmd", "a")
main_slides:write(end_content)
main_slides:close()
```

Is there a way to refer to the currently rendered file, instead of hardcoding its name into the script (e.g. "template.qmd")? Ideally, I would like to call the lua script as a fragment in the YAML. Thanks a lot!

**Comment by baptiste** [comment]
I'm not sure this is the right approach: Lua filters are called relatively late in the rendering process, after Quarto has already processed the qmd files and pandoc has parsed the results into an abstract syntax tree. (Pandoc Lua readers act earlier than that, but I doubt they're relevant here). So in that sense, it seems too late to include a qmd file at this stage.

It sounds like you could use [quarto includes](https://quarto.org/docs/authoring/includes.html) instead? Or perhaps adjust the document template, if using latex/html/typst/etc formats.

**Reply by alesvomacka** [reply]
Thanks. I have thought about the include shortcode, but I couldn't think of a way to use it without me having to manually adding it to the document.

By adjusting template, you mean adjusting the (for example) reveal.js template itself? If so, could you point me to a document that describes the process? I'm not familiar with javascript to know where to start.

**Reply by baptiste** [reply]
You should share a minimal reproducible example of what you're trying to do (main slides and document to insert), but also take a look at these templates:

https://quarto.org/docs/journals/templates.html#revealjs-partials

You could copy the template file and add your custom slide after [`$body$`](https://github.com/quarto-dev/quarto-cli/blob/main/src/resources/formats/revealjs/pandoc/template.html#L58), or something similar.

**Comment by cderv** [comment]
You could also consider Quarto Project, and using a pre-render script ([About Project Script](https://quarto.org/docs/projects/scripts.html#pre-and-post-render). You could add your content using a script to all the other .qmd that requires it before Quarto does its processing 

> It sounds like you could use [quarto includes](https://quarto.org/docs/authoring/includes.html) instead?

Using `includes` shortcode is really just one line, and this is the best way to add the content of a .qmd to  another .qmd and have the merged document process as a single `.qmd`

Using Lua, you could just consider the `Doc` block, and add your parsed `ending.qmd` to it (See `pandoc.read()`). This will add content to the already parsed AST, so you don't need the document name. However, this will work more with `ending.md` than a `ending.qmd` (as the content should be only Pandoc processed content - no code chunk, or complex Quarto representation).

> You could copy the template file and add your custom slide after [$body$](https://github.com/quarto-dev/quarto-cli/blob/main/src/resources/formats/revealjs/pandoc/template.html#L58), or something similar.

If the aim is to add some content after body, then using the Pandoc includes features is probably easier than modifying template 
https://quarto.org/docs/output-formats/html-basics.html#includes

Hope it helps

**Comment by cl-roberts** [comment, marked as answer]
Hi there,

I made a filter called [append-qmd](https://github.com/clroberts-adfg/lua-filters/tree/main/append-qmd) that I believe does what you're after. Simply acquire that repository and add your contact info to a `.qmd` file; place it in the same directory as the `.lua` script. 

Also, pandoc has a filter called [include-files](https://github.com/pandoc/lua-filters/tree/master/include-files) that is a little more flexible but is less automated. You have to write a code block with class `.include` in the body of your main document to prompt inclusion of external markdown files .

**Reply by mcanouil** [reply]
I am puzzled by the whole idea here ...
To avoid manually adding an include shortcode, the idea is to manually add a filter that does the include/append, isn't it the same thing but more complicated?

**Reply by cscheid** [reply]
@mcanouil Remember that filter inclusion can happen in [`_quarto.yml` or on a per-directory basis with `_metadata.yml`](https://quarto.org/docs/projects/quarto-projects.html#shared-metadata), which means you only need to do it once per project/directory.

**Reply by mcanouil** [reply]
Oh yes, that's true.
I tend to do one repo = one thing 😅
So, if the goal is to have a "slides/presentations mono-repo" that makes total sense.

**Reply by cl-roberts** [reply]
I agree that an `include` shortcode is a better approach in most cases. In my particular use case, I have a handful of related documents in different formats within a project and wanted to add generic appendices to all using a shared YAML, so I think this approach makes sense. I could also envision a scenario where you need to append many files with a shell script, and adding a line to the YAML is easier then adding shortcode.

**Reply by cderv** [reply]
@clroberts-adfg Nice filter! 
does your filter works with any content in the `.qmd` ? Which format are you reading ? I see `pandoc.read(..., nil)` - what is `nil` referring to as reading format ? 

@cscheid this type of filter is using `pandoc.read()`. I believe in Quarto we are using a specific `.qmd` reader for `.qmd` file. Could there be some side effect, like feature not working if the filter is just using `pandoc.read()` without using this specific reader ? 

Overall, using external Lua filter for that seems an approach that would work for `.md` file mainly, without any .qmd specific content. 
I wonder if there should be a `quarto.read()` exported in our Quarto API. Maybe too complex though...

**Reply by cscheid** [reply]
@cderv You're correct, if the code uses `pandoc.read()`, then (for example), triple-trick code syntax is unlikely to work, as are shortcodes. You are also correct that we need a `quarto.read()` API or something like that. But we don't have it yet.

**Reply by cderv** [reply]
Thanks for confirming !

**Reply by cl-roberts** [reply]
@cderv That's a good point, this is really appending `.qmd` files with `.md` files since `.qmd` content [can't be read](https://pandoc.org/chunkedhtml-demo/3.1-general-options.html) as far as I know. The `nil` in `pandoc.read` just specifies the default format value (which is markdown); I changed it to be more explicit. I think a quarto-specific reader would be fantastic; otherwise I've just included plots as `.png` files, tables as pipe tables, etc.

**Comment by alesvomacka** [comment]
@clroberts-adfg  This is pretty much exactly what I was after. Thanks a lot!
